Focus on ready to use, remove maturity labels with industry misalignment (#729)
publiccodenet/standard: criteria/document-maturity.md

## Requirements

* A codebase MUST be versioned.

> * A codebase version that is ready to use MUST only depend on other codebase versions that are also ready to use.

Reviewer: Unless I misunderstand, if the point is to mention external dependencies (as opposed to self-dependencies), then this would be clearer to me in plural.

```suggestion
* A codebase version that is ready to use MUST only depend on other codebases versions that are also ready to use.
```

* A codebase which has no versions ready for use MUST document that clearly.

> * A codebase version that is not yet ready to use MUST use semantic versioning with a 0 for the major version number or have one of the labels: pre-alpha, alpha, beta or release candidate.

Reviewer: ```suggestion
* A codebase version that is not yet ready to use MUST use semantic versioning with [major version number `0`](https://semver.org/#spec-item-4), or have one of the following labels: `pre-alpha`, `alpha`, `beta` or `release candidate`.
```

> * A codebase version that is not yet ready to use MUST use semantic versioning with a 0 for the major version number or have one of the labels: pre-alpha, alpha, beta or release candidate.

Reviewer: Is this “label” in the sense of [SemVer§9](https://semver.org/#spec-item-9) or independent from the release / packaging system? The current phrasing is ambiguous and I initially understood it as SemVer§9 until I read label `release candidate` with a space, which would not be a valid SemVer label.

Author: Today, this is not [SemVer§9](https://semver.org/#spec-item-9). One can name the release separate from the semver tag, consider https://github.com/publiccodenet/standard/releases/ which so far have labels of the form "tenth update" and version tag of `0.3.0`.

That said, it would clearly be much, much better if they could be exactly the same. Thus, 'release candidate' creates a problem. This should absolutely be addressed before merging.

> * A codebase version that is not yet ready to use MUST use semantic versioning with a 0 for the major version number or have one of the labels: pre-alpha, alpha, beta or release candidate.

Reviewer: I am not sure I understand the added value of mandating specific labels, unless they match specific phases that are mandated by the Standard, in which case a link would seem appropriate to me. Some projects like to use `rc` (for “release candidate”), some others “golden build”, “canary”, “discovery”… On top, some package managers will not allow spaces in labels (for example, tags in NPM, IIRC), and `release candidate` as currently written seems to be a spec item.

My suggestion for solving most comments I leave here would be to stick to SemVer.

Author: I would strongly encourage most codebases to stick to SemVer, but there are codebases for which this does not make sense. One examples would be a codebase for a website which automatically deploys from `develop` to `live` if the tests are green, another would be a rolling update linux distribution. These are, however, the exceptions. SemVer is well suited to most codebases.

Reviewer 2: > I am not sure I understand the added value of mandating specific labels, unless they match specific phases that are mandated by the Standard, in which case a link would seem appropriate to me.

I think this part was an insightful observation. Let's think a bit more about what is really adding value to those who come looking at a codebase.

Author: The current version of this PR no longer mandates specific labels.

* A codebase SHOULD contain a log of changes from version to version, for example in the `CHANGELOG`.

> * It is OPTIONAL to use semantic versioning.

Reviewer: This reads inconsistent with line 16. At least, make things clearer.
```suggestion
* It is OPTIONAL to use semantic versioning for versions that are ready to use.
```

However, I find this inconsistent: with this ruleset, I am allowed to use `0` as a major version number or have one of the labels listed on line 16 with a version that is ready to use, since I am not bound by semver anymore 🤯

Author: I do not think this is correct. The codebase must be versioned. If the codebase is versioned using semantic versioning, I would expect that versioning policy to apply to all releases, regardless of -beta or whatever.

Author: If a codebase is using semver rules, then the `0.x.y` may not be a `beta` but it is clear that it is not ready to use. Also, 1.x.y-beta-4 is clear that it is not ready to use.

A release with a label like 1.x.y would be expected to be ready to use unless it is clearly labelled otherwise.

## Why this is important

> Clearly signalling a codebase's maturity helps others decide whether to reuse, invest in or contribute to it. Clearly labeling the maturity of versions helps others decide which version to use and the level of testing required.

Reviewer: ```suggestion
Clearly signalling a codebase's maturity helps others decide whether to reuse, invest in or contribute to it. Clearly labelling the maturity of versions helps others decide which version to use, how and when to upgrade, and the level of testing required.
```

> Clearly signalling a codebase's maturity helps others decide whether to reuse, invest in or contribute to it. Clearly labeling the maturity of versions helps others decide which version to use and the level of testing required.

Reviewer: > I am not certain that the “level of testing required” can be hinted at by the producer: each reuse might have different QA & certification requirements. This is fine by me, but I don't find that part necessary.

## Management: what you need to do

> * Make sure that services only rely on codebase versions of equal or greater maturity than the service. For example, don't use a beta version of a codebase in a production service or an alpha codebase version in a beta service.

Reviewer: ```suggestion
* Make sure that services only rely on codebases versions of equal or greater maturity than the service. For example, don't use a beta version of a codebase in a production service or an alpha version codebase in a beta service.
```

* If the codebase version is not ready for general use, work with the developers to ensure the version is properly labeled:

>   * pre-alpha: to internally test or explore technical possibilities,
>   * alpha: to do guided tests with a limited set of users, typically not feature complete,
>   * beta: to open up testing to a larger section of the general public, for example to test if the codebase works at scale, typically feature complete,
>   * release candidate: code that is ready to be released but has not completed release testing.

Reviewer: ```suggestion
  * `pre-alpha`: to internally test or explore technical possibilities;
  * `alpha`: to do guided tests with a limited set of users, typically not feature complete;
  * `beta`: to open up testing to a larger section of the general public, for example to test if the codebase works at scale, typically feature complete;
  * `release candidate`: code that is ready to be released but has not completed release testing.
```
